### DISCUSSION TITLE
Allowing Lambda Expressions with Statement Bodies Describing Iterators to be Converted to Expression Trees

### DISCUSSION THREAD

**Discussion by AdamSobieski** [dotnet/csharplang#3094] [Language Ideas]
Building on issue https://github.com/dotnet/csharplang/issues/3093, I would like to open an issue for discussing allowing lambda expressions with statement bodies describing iterators to be able to be converted to expression trees in C#.

```cs
Expression<Func<IEnumerable<int>>> expr = () =>
{
    for (int index = 0; index < 100; ++index)
    {
        yield return index;
    }
};
```

It would be quite useful if the System.Linq.Expressions expression tree model could represent the iterator pattern and if the `Compile()` method could, at runtime, generate iterators.

Pertinent to this issue are [CS0834](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/cs0834) (a lambda expression with a statement body cannot be converted to an expression tree) and [CS1621](https://docs.microsoft.com/en-us/dotnet/csharp/misc/cs1621) (the yield statement cannot be used inside an anonymous method or lambda expression).

**Comment by Joe4evr** [comment]
This falls under #158.
